sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java

```diff
@@ -943,36 +936,57 @@ public Mono<VerifyResult> verifyData(SignatureAlgorithm algorithm, byte[] data,
 
     private Mono<Boolean> isValidKeyLocallyAvailable() {
         if (localOperationNotSupported) {
-            return Mono.just(false);
+            return Mono.defer(() -> Mono.just(false));
+        } else if (this.key == null && implClient.getKeyCollection() != null) {
+            // Get key from service and validate it. Then attempt to create a local cryptography client or default to
+            // service.
+            Mono<JsonWebKey> jsonWebKeyMono = CryptographyUtils.SECRETS_COLLECTION.equals(implClient.getKeyCollection())
+                ? implClient.getSecretKeyAsync() : getKey().map(KeyVaultKey::getKey);
+
+            return jsonWebKeyMono
+                .map(this::validateJsonWebKeyAndCreateLocalClient)
+                .onErrorResume(HttpResponseException.class, this::handleKeyRetrievalError);
+        } else {
+            return Mono.defer(() -> Mono.just(true));
         }
+    }
+
+    private boolean validateJsonWebKeyAndCreateLocalClient(JsonWebKey jsonWebKey) {
+        key = jsonWebKey;
+
+        if (jsonWebKey.isValid()) {
+            // Create a local key cryptography client if one has not been created yet.
+            if (localKeyCryptographyClient == null) {
+                try {
+                    localKeyCryptographyClient = initializeCryptoClient(jsonWebKey, implClient, LOGGER);
+                } catch (RuntimeException e) {
+                    localOperationNotSupported = true;
+
+                    LOGGER.warning("Defaulting to service use for cryptographic operations.", e);
 
-        if (key == null && implClient.getKeyCollection() != null) {
-            Mono<JsonWebKey> jsonWebKeyMono = CryptographyUtils.SECRETS_COLLECTION.equals(implClient.getKeyCollection())
-                ? getSecretKey()
-                : getKey().map(KeyVaultKey::getKey);
-
-            return jsonWebKeyMono.map(jsonWebKey -> {
-                key = jsonWebKey;
-
-                if (jsonWebKey.isValid()) {
-                    if (localKeyCryptographyClient == null) {
-                        try {
-                            localKeyCryptographyClient = initializeCryptoClient(jsonWebKey, implClient, LOGGER);
-                        } catch (RuntimeException e) {
-                            localOperationNotSupported = true;
-                            LOGGER.warning("Defaulting to service use for cryptographic operations.", e);
-
-                            return false;
-                        }
-                    }
-
-                    return true;
-                } else {
                     return false;
                 }
-            });
+            }
+
+            // Valid key is locally available and local client was created.
+            return true;
         } else {
-            return Mono.just(true);
+            return false;
+        }
+    }
+
+    private Mono<Boolean> handleKeyRetrievalError(HttpResponseException e) {
```

Comment:
This cannot be static because it updates the instance variable `localOperationSupported`.

Reply:
Also, we set things the way they are to avoid blocking anywhere in these calls and keep things truly async, so while this can return a `boolean` instead of a `Mono<Boolean>`, we'll eventually have to wrap that value in a `Mono` at some point. I think it's just a matter of where we create said Mono. I don't mind making the entry point wrap things to be able to reuse some of this logic in the sync client.